pkg/gds/admin.go

```diff
@@ -1006,6 +1007,133 @@ func (s *Admin) Resend(c *gin.Context) {
 	c.JSON(http.StatusOK, out)
 }
 
+// ReviewTimeline returns a list of time series records containing registration state counts by week.
+func (s *Admin) ReviewTimeline(c *gin.Context) {
+	// Go needs this constant to determine the time format
+	const timeFormat = "2006-01-02"
+	var (
+		err          error
+		in           *admin.ReviewTimelineParams
+		out          *admin.ReviewTimelineReply
+		startTime    time.Time
+		endTime      time.Time
+		earliestTime time.Time
+		weekTime     time.Time
+		numWeeks     int
+		vaspCounts   []map[string]bool
+	)
+
+	// Get request parameters
+	in = new(admin.ReviewTimelineParams)
+	if err = c.ShouldBindQuery(&in); err != nil {
+		log.Warn().Err(err).Msg("could not bind request with query params")
+		c.JSON(http.StatusBadRequest, admin.ErrorResponse(err))
+		return
+	}
+
+	// Default value for start date
+	earliestTime = time.Date(2020, time.January, 1, 0, 0, 0, 0, time.Local)
+	if in.Start != "" {
+		// Parse start date
+		if startTime, err = time.Parse(timeFormat, in.Start); err != nil {
+			log.Warn().Err(err).Msg("could not parse start date")
+			c.JSON(http.StatusBadRequest, admin.ErrorResponse(fmt.Errorf("invalid start date: %s", in.Start)))
+		}
+		// Hard limit on the earliest date to avoid making the server do unnecessary work
+		if startTime.Before(earliestTime) {
+			startTime = earliestTime
+		}
+	} else {
+		startTime = earliestTime
+	}
+
+	if in.End != "" {
+		// Parse end date
+		if endTime, err = time.Parse(timeFormat, in.End); err != nil {
+			log.Warn().Err(err).Msg("could not parse end date")
+			c.JSON(http.StatusBadRequest, admin.ErrorResponse(fmt.Errorf("invalid end date: %s", in.End)))
+		}
+	} else {
+		// Default value for end date
+		endTime = time.Now()
+	}
+	if startTime.After(endTime) {
+		log.Warn().Err(fmt.Errorf("start date after end date")).Msg("invalid timeline request: start date can't be after current date")
+		c.JSON(http.StatusBadRequest, admin.ErrorResponse(fmt.Errorf("start date must be before end date")))
+	}
+
+	// Initialize required counting structs
+	numWeeks = int(endTime.Sub(startTime).Hours()/24/7) + 1
+	vaspCounts = make([]map[string]bool, numWeeks)
+	out = &admin.ReviewTimelineReply{
+		Weeks: make([]admin.ReviewTimelineRecord, numWeeks),
+	}
+	weekTime = startTime
+	for i := 0; i < numWeeks; i++ {
+		weekDate := weekTime.Format(timeFormat)
+		record := admin.ReviewTimelineRecord{
+			Week:          weekDate,
+			VASPsCount:    0,
+			Registrations: make(map[string]int),
+		}
+		var s int32
+		for s = 0; s <= int32(pb.VerificationState_ERRORED); s++ {
+			record.Registrations[pb.VerificationState_name[s]] = 0
+		}
```

Comment:
Just to clarify - the default value of the map is already 0 so if you access a key that doesn't exist in the map it will return 0 without an error. Addition also works, e.g.

```go
counts := make(map[string]int)
counts["foo"]++
counts["bar"]++
counts["other"] += 11
fmt.Println(counts)
```

Prints: `map[bar:1 foo:1 other:11]`

Therefore this section is not necessary for initialization, but is necessary to ensure that all states are represented with a default count of 0  (otherwise the JSON would simply omit the state) - is that correct? If so, I think a comment to that effect would be very helpful to prevent well-intending future contributors from removing this section. 

Alternatively we could talk to @elysee15 to see if he can handle a "missing" state with a count of 0 - this would reduce the amount of data transferred, only sending states with counts > 0. That would be better, but only if the front-end graph can handle that easily.

Reply:
I think that could work, depending on how easy it is to parse on the front-end. @elysee15 Does the front-end know about the possible verification states?